contributors/design-proposals/preserve-order-in-strategic-merge-patch.md

```diff
@@ -0,0 +1,547 @@
+
+The patched list will look like:
+
+```
+mergingList:
+- parallelListItem1 \
+  ...                |===> items from the parallel list
+- parallelListItemN /
+- otherItem1 \
+  ...         |===> items in the server's list but not in the parallel list
+- otherItemN /
+```
+
+### When $setElementOrder is not present and patching a list
+
+The new directive $setElementOrder is optional.
+The new server applying a patch request without the directive will be
+a best-effort operation to keep the order of elements that are already in the list:
+
+For changes and deletions, the server will just merge the change without sorting them.
+So the items in the list will retain the order.
+
+For additions, the items will be appended to the end of the list.
+
+So it changes a little from previous releases, but is still fully backward compatible with old clients.
+
+Examples where A and C have been changed, B has been deleted and D has been added.
+
+Patch:
+
+```yaml
+list:
+- A'
+- $patch: delete
+  B
+- D
+```
+
+Live:
+
+```yaml
+list:
+- C
+- B
+- A
+```
+
+Result:
+
+```yaml
+list:
+- C
+- A
+- D
+```
+
+### `$setElementOrder` may contain elements not present in the patch list
+
+The $setElementOrder value may contain elements that are not present in the patch
+but present in the list to be merge to reorder the elements as part of the merge.
+
+Example where A & B have not changed:
+
+Patch:
+
+```yaml
+$setElementOrder/list:
+- A
+- B
+```
+
+Live:
+
+```yaml
+list:
+- B
+- A
+```
+
+Result:
+
+```yaml
+list:
+- A
+- B
+```
+
+### When the list to be merged contains elements not found in `$setElementOrder`
+
+If the list to be merged contains elements not found in $setElementOrder,
+they will come after all elements defined in $setElementOrder, but keep their relative ordering.
+
+Example where A & B have been changed:
+
+```yaml
+$setElementOrder/list:
+- A
+- B
+list:
+- A
+- B
+```
+
+```yaml
+list:
+- C
+- B
+- D
+- A
+- E
+```
+
+Result:
+
+```yaml
+list:
+- A
+- B
+- C
+- D
+- E
+```
+
+### When `$setElementOrder` contains elements not found in the list to be merged
+
+If `$setElementOrder` contains elements not found in the list to be merged,
+the elements that are not found will be ignored instead of failing the request.
+
+Patch:
+```yaml
+$setElementOrder/list:
+- C
+- A
+- B
+list:
+- A
+- B
+```
+
+Live:
+```yaml
+list:
+- A
+- B
+```
+
+Result:
+
+```yaml
+list:
+- A
+- B
+```
+
+## Version Skew and Backwards Compatibility
+
+The new version patch is always a superset of the old version patch.
+The new patch has one additional parallel list which will be dropped by the old server.
+
+As mentioned [above](#when-setelementorder-is-not-present-and-patching-a-list),
+the new directive is optional.
+Patch requests without the directive will change a little,
+but still be fully backward compatible.
+
+### kubectl
+If an old kubectl sends a old patch to a new server, the server will not preserve the order which behave as an old server.
```

Comment:
@liggitt I was thinking about this as well.  

I think you are saying, can you just use the ordering in the patch list instead of sending another directive.

The issue here is: If you have a list of [a,b,c] and want to patch b you would either send a patch [b] or send a patch [a,b,c].  The former lacks the ordering, and the latter will recreate [a,c] if they were deleted from the server.